In [1]:
!mkdir ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json
!kaggle datasets download -d mlg-ulb/creditcardfraud
!unzip creditcardfraud.zip
# this was to load the dataset into our consideration of files for this notebook

cp: cannot stat 'kaggle.json': No such file or directory
chmod: cannot access '/root/.kaggle/kaggle.json': No such file or directory
Dataset URL: https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud
License(s): DbCL-1.0
 86% 57.0M/66.0M [00:00<00:00, 74.7MB/s]
100% 66.0M/66.0M [00:00<00:00, 72.6MB/s]
Archive:  creditcardfraud.zip
  inflating: creditcard.csv          


In [2]:
import pandas as pd
# this kinda helps in reading through or accessing info from csv type tabular files
# similar things were directly used in R
df = pd.read_csv('creditcard.csv')
# to read the csv file
df

Time         V1         V2        V3        V4        V5  \
0            0.0  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1            0.0   1.191857   0.266151  0.166480  0.448154  0.060018   
2            1.0  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3            1.0  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4            2.0  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  149.62   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724    2.69   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  378.66   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  123.50   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153   69.99   
...          ...       ...       ...       ...       ...       ...     ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [3]:
df['Class'].value_counts()
# this tells within that column, Class here
# what are the different values and how many times they occur
# this helps us in knowing positive and negative classes, their frequency
# useful in classification like here

Class
0    284315
1       492
Name: count, dtype: int64

In [4]:
#df.hist(bins = 30, figsize = (30,30))
# trying this hist function more to understand it
df.hist(bins = 20, figsize = (20,20))
# bins = no. of intervals
#figsize = dimensions

array([[<Axes: title={'center': 'Time'}>, <Axes: title={'center': 'V1'}>,
        <Axes: title={'center': 'V2'}>, <Axes: title={'center': 'V3'}>,
        <Axes: title={'center': 'V4'}>, <Axes: title={'center': 'V5'}>],
       [<Axes: title={'center': 'V6'}>, <Axes: title={'center': 'V7'}>,
        <Axes: title={'center': 'V8'}>, <Axes: title={'center': 'V9'}>,
        <Axes: title={'center': 'V10'}>, <Axes: title={'center': 'V11'}>],
       [<Axes: title={'center': 'V12'}>, <Axes: title={'center': 'V13'}>,
        <Axes: title={'center': 'V14'}>, <Axes: title={'center': 'V15'}>,
        <Axes: title={'center': 'V16'}>, <Axes: title={'center': 'V17'}>],
       [<Axes: title={'center': 'V18'}>, <Axes: title={'center': 'V19'}>,
        <Axes: title={'center': 'V20'}>, <Axes: title={'center': 'V21'}>,
        <Axes: title={'center': 'V22'}>, <Axes: title={'center': 'V23'}>],
       [<Axes: title={'center': 'V24'}>, <Axes: title={'center': 'V25'}>,
        <Axes: title={'center': 'V26'}>, <

Error in callback <function flush_figures at 0x789728b23a30> (for post_execute):


KeyboardInterrupt: 

In [5]:
df.describe()
# summary of the dataset

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [6]:
from sklearn.preprocessing import RobustScaler
# scikit learn
df_2 = df.copy()
#copy the database into df_2 so original data is not affected
df_2['Amount'] = RobustScaler().fit_transform(df_2['Amount'].to_numpy().reshape(-1,1))
# replace the original values in the 'Amount' column with their scaled values
# Converts the 'Amount' column to a NumPy array.
# Reshapes the NumPy array to a 2D array with one column and as many rows as necessary (the -1 allows the rows to be determined automatically)
# Fits the RobustScaler to the data and transforms it in one step
#df_2['Amount'].hist()
time = df_2['Time']
df_2['Time'] = (time - time.min())/(time.max() - time.min())
df_2

Time         V1         V2        V3        V4        V5  \
0       0.000000  -1.359807  -0.072781  2.536347  1.378155 -0.338321   
1       0.000000   1.191857   0.266151  0.166480  0.448154  0.060018   
2       0.000006  -1.358354  -1.340163  1.773209  0.379780 -0.503198   
3       0.000006  -0.966272  -0.185226  1.792993 -0.863291 -0.010309   
4       0.000012  -1.158233   0.877737  1.548718  0.403034 -0.407193   
...          ...        ...        ...       ...       ...       ...   
284802  0.999965 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  0.999971  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  0.999977   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  0.999977  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  1.000000  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
0       0.462388  0.239599  0.098698  0.363787  ... -0.018307  0.277838   
1      -0.082361 -0.078803  0.085102 -0.255425  ... -0.225775 -0.638672   
2       1.800499  0.791461  0.247676 -1.514654  ...  0.247998  0.771679   
3       1.247203  0.237609  0.377436 -1.387024  ... -0.108300  0.005274   
4       0.095921  0.592941 -0.270533  0.817739  ... -0.009431  0.798278   
...          ...       ...       ...       ...  ...       ...       ...   
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28    Amount  \
0      -0.110474  0.066928  0.128539 -0.189115  0.133558 -0.021053  1.783274   
1       0.101288 -0.339846  0.167170  0.125895 -0.008983  0.014724 -0.269825   
2       0.909412 -0.689281 -0.327642 -0.139097 -0.055353 -0.059752  4.983721   
3      -0.190321 -1.175575  0.647376 -0.221929  0.062723  0.061458  1.418291   
4      -0.137458  0.141267 -0.206010  0.502292  0.219422  0.215153  0.670579   
...          ...       ...       ...       ...       ...       ...       ...   
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731 -0.296653   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527  0.038986   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561  0.641096   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533 -0.167680   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  2.724796   

        Class  
0           0  
1           0  
2           0  
3           0  
4           0  
...       ...  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[284807 rows x 31 columns]

In [7]:
df_2 = df_2.sample(frac=1, random_state=1)
df_2
# randomly shuffled order
# frac=1: Specifies that you want to sample 100% of the rows
# random_state=1: Sets a seed for the random number generator, ensuring that the shuffling is reproducible. Using the same random_state value will always produce the same shuffled DataFrame when you run the code again


Time        V1        V2        V3        V4        V5        V6  \
169876  0.693938 -0.611712 -0.769705 -0.149759 -0.224877  2.028577 -2.019887   
127467  0.453377 -0.814682  1.319219  1.329415  0.027273 -0.284871 -0.653985   
137900  0.476770 -0.318193  1.118618  0.969864 -0.127052  0.569563 -0.532484   
21513   0.183556 -1.328271  1.018378  1.775426 -1.574193 -0.117696 -0.457733   
134700  0.468326  1.276712  0.617120 -0.578014  0.879173  0.061706 -1.472002   
...          ...       ...       ...       ...       ...       ...       ...   
21440   0.183261 -2.986845 -8.663978 -1.910863  0.664058 -3.934875  0.861269   
117583  0.432480  0.937083 -0.849673  0.524186 -0.020031 -0.606327  0.692302   
73349   0.318852 -1.149963  1.696462  1.637114  2.658991 -0.021502  0.192287   
267336  0.941757  1.754554 -0.699398 -0.076332  0.443915 -0.672082  0.389061   
128037  0.454743 -0.707635  0.493302  2.648089  1.064807 -0.680271  1.183838   

              V7        V8        V9  ...       V21       V22       V23  \
169876  0.292491 -0.523020  0.358468  ... -0.075208  0.045536  0.380739   
127467  0.321552  0.435975 -0.704298  ... -0.128619 -0.368565  0.090660   
137900  0.706252 -0.064966 -0.463271  ... -0.305402 -0.774704 -0.123884   
21513   0.681867 -0.031641  0.383872  ... -0.220815 -0.419013 -0.239197   
134700  0.373692 -0.287204 -0.084482  ... -0.160161 -0.430404 -0.076738   
...          ...       ...       ...  ...       ...       ...       ...   
21440   1.647511 -0.480963 -1.546866  ...  1.252092 -0.993085 -2.173147   
117583 -0.463724  0.148857  0.785062  ... -0.143322 -0.479981 -0.237902   
73349   0.205204  0.588754 -1.187820  ...  0.025147  0.086506 -0.262748   
267336 -0.807534  0.202915  0.858635  ...  0.141950  0.358412  0.259748   
128037  0.169413  0.074553  1.247988  ... -0.102350  0.323975 -0.172601   

             V24       V25       V26       V27       V28     Amount  Class  
169876  0.023440 -2.220686 -0.201146  0.066501  0.221180  -0.282401      0  
127467  0.401147 -0.261034  0.080621  0.162427  0.059456  -0.279746      0  
137900 -0.495687 -0.018148  0.121679  0.249050  0.092516  -0.294977      0  
21513   0.009967  0.232829  0.814177  0.098797 -0.004273  -0.084119      0  
134700  0.258708  0.552170  0.370701 -0.034255  0.041709  -0.296793      0  
...          ...       ...       ...       ...       ...        ...    ...  
21440   0.145570 -0.235062 -0.227411 -0.382702  0.404045  32.002515      0  
117583 -0.715247  0.251418  0.975406 -0.060168  0.023771   2.086495      0  
73349   0.321538  0.341667  0.210343 -0.162047  0.031193  -0.201495      0  
267336  0.746839 -0.560808  0.104636 -0.005853 -0.019622   1.017257      0  
128037  0.126965 -0.001998 -0.398741 -0.385589 -0.205589   0.500245      0  

[284807 rows x 31 columns]

In [8]:
train, test, val = df_2[:240000], df_2[240000:262000], df_2[262000:]
# here we spilt the dataset effectively
train['Class'].value_counts(), test['Class'].value_counts(), val['Class'].value_counts()
# we find how many positive and negative classes are in each

(Class
 0    239589
 1       411
 Name: count, dtype: int64,
 Class
 0    21955
 1       45
 Name: count, dtype: int64,
 Class
 0    22771
 1       36
 Name: count, dtype: int64)

In [9]:
train_np, test_np, val_np = train.to_numpy(), test.to_numpy(), val.to_numpy()
train_np.shape, test_np.shape, val_np.shape
# conversion to numpy arrays for each split dataset

((240000, 31), (22000, 31), (22807, 31))

In [10]:
# x -> input, y -> output

x_train, y_train = train_np[:, :-1], train_np[:, -1]
# -1 refers to last column here input has everything except it,
# output has only that, this for every row
x_test, y_test = test_np[:, :-1], test_np[:, -1]
x_val, y_val = val_np[:, :-1], val_np[:, -1]
#same for test and val set
x_train.shape, y_train.shape, x_test.shape, y_test.shape, x_val.shape, y_val.shape

((240000, 30), (240000,), (22000, 30), (22000,), (22807, 30), (22807,))

In [13]:
# train a logistic regression model on the training data and evaluate its performance

from sklearn.linear_model import LogisticRegression
logistic_model = LogisticRegression()
# instance of the LogisticRegression model with default parameters.
logistic_model.fit(x_train, y_train)
# trains the logistic regression model
logistic_model.score(x_train, y_train)
# returns the mean accuracy
# accuracy score is the ratio of correctly predicted
# instances to the total instances in the training set

/usr/local/lib/python3.10/dist-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.9992375

In [14]:
# evaluate the performance of the
# logistic regression model on the validation data

from sklearn.metrics import classification_report
y_val_pred = logistic_model.predict(x_val)
# predictions on the validation set
print(classification_report(y_val, y_val_pred, target_names=['Not Fraud', 'Fraud']))
# if 0 Not Fraud and 1 Fraud, target names our convinience
# comparison is done



              precision    recall  f1-score   support

   Not Fraud       1.00      1.00      1.00     22771
       Fraud       0.73      0.53      0.61        36

    accuracy                           1.00     22807
   macro avg       0.87      0.76      0.81     22807
weighted avg       1.00      1.00      1.00     22807



Precision: The ratio of true positive predictions to the total number of positive predictions (true positives + false positives). It indicates how many of the predicted positive instances are actually positive.

Recall: The ratio of true positive predictions to the total number of actual positive instances (true positives + false negatives). It indicates how many of the actual positive instances are correctly predicted.

F1-Score: The harmonic mean of precision and recall. It provides a single metric that balances the two.

Support: The number of actual occurrences of each class in the validation set.

understand what are false negatives and false positives

The macro average is the unweighted mean of the metrics, treating all classes equally, while the weighted average takes class support into account.

In [14]:
# shallow neural network using the TensorFlow Keras API

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import InputLayer, Dense, BatchNormalization
from tensorflow.keras.callbacks import ModelCheckpoint
# Import Necessary Modules

shallow_nn = Sequential()
# Initialize the Sequential Model
shallow_nn.add(InputLayer((x_train.shape[1],)))
# add input layer, number of features in the training data, here 30
shallow_nn.add(Dense(2, 'relu'))
# This adds a dense (fully connected) layer with
# 2 units and ReLU activation function.
shallow_nn.add(BatchNormalization())
# This normalizes the activations of the previous layer at each batch,
# which helps with training stability and speed.
shallow_nn.add(Dense(1, 'sigmoid'))
# This adds the output layer with 1 unit and a sigmoid activation
# function, suitable for binary classification tasks.

checkpoint = ModelCheckpoint('shallow_nn', save_best_only=True)
# This callback saves the model with the
# best validation accuracy during training.
shallow_nn.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
# This specifies the Adam optimizer,
# binary cross-entropy loss function, and
# accuracy as the metric to monitor.



In [15]:
shallow_nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 2)                 62        
                                                                 
 batch_normalization (Batch  (None, 2)                 8         
 Normalization)                                                  
                                                                 
 dense_1 (Dense)             (None, 1)                 3         
                                                                 
Total params: 73 (292.00 Byte)
Trainable params: 69 (276.00 Byte)
Non-trainable params: 4 (16.00 Byte)
_________________________________________________________________


In [16]:
shallow_nn.fit(x_train, y_train, validation_data=(x_val, y_val), epochs=5, callbacks=checkpoint)

Epoch 1/5
7500/7500 [==============================] - 30s 4ms/step - loss: 0.0479 - accuracy: 0.9894 - val_loss: 0.0075 - val_accuracy: 0.9991
Epoch 2/5
7500/7500 [==============================] - 26s 4ms/step - loss: 0.0035 - accuracy: 0.9993 - val_loss: 0.0072 - val_accuracy: 0.9991
Epoch 3/5
7500/7500 [==============================] - 26s 3ms/step - loss: 0.0033 - accuracy: 0.9994 - val_loss: 0.0088 - val_accuracy: 0.9990
Epoch 4/5
7500/7500 [==============================] - 26s 4ms/step - loss: 0.0033 - accuracy: 0.9993 - val_loss: 0.0091 - val_accuracy: 0.9991
Epoch 5/5
7500/7500 [==============================] - 28s 4ms/step - loss: 0.0032 - accuracy: 0.9994 - val_loss: 0.0102 - val_accuracy: 0.9989


The training process will run for 5 epochs, and the model checkpoint callback will save the best version of the model based on the validation accuracy.

fit method: This method trains the model on the training data for a specified number of epochs and validates it on the validation data.

epochs=5: The number of complete passes through the training dataset.

callbacks=[checkpoint]: A list of callback functions to apply during training, in this case, saving the best model based on validation accuracy.



In [25]:
def neural_net_predictions(model, x):
  return (model.predict(x).flatten() > 0.5).astype(int);
neural_net_predictions(shallow_nn, x_val)

713/713 [==============================] - 2s 2ms/step


array([0, 0, 0, ..., 0, 0, 0])

In [26]:
print(classification_report(y_val, neural_net_predictions(shallow_nn, x_val), target_names=['Not Fraud', 'Fraud']))

713/713 [==============================] - 1s 2ms/step
              precision    recall  f1-score   support

   Not Fraud       1.00      1.00      1.00     22771
       Fraud       0.61      0.78      0.68        36

    accuracy                           1.00     22807
   macro avg       0.80      0.89      0.84     22807
weighted avg       1.00      1.00      1.00     22807



In [27]:
# Random Forest model on the training data

from sklearn.ensemble import RandomForestClassifier
# Import Necessary Modules
rf = RandomForestClassifier(max_depth=2, n_jobs=-1)
# Initialize and Train the Random Forest Classifier
rf.fit(x_train, y_train)
print(classification_report(y_val, rf.predict(x_val), target_names=['Not Fraud', 'Fraud']))
# max_depth=2: Limits the maximum depth of each tree in the forest, which can help prevent overfitting.
# n_jobs: This parameter specifies the number of CPU cores to use for the computatio

              precision    recall  f1-score   support

   Not Fraud       1.00      1.00      1.00     22771
       Fraud       0.81      0.47      0.60        36

    accuracy                           1.00     22807
   macro avg       0.90      0.74      0.80     22807
weighted avg       1.00      1.00      1.00     22807



In [30]:
# train a Gradient Boosting model on the training data

from sklearn.ensemble import GradientBoostingClassifier
gbc = GradientBoostingClassifier(n_estimators=50, learning_rate=1.0, max_depth=1, random_state=3)
gbc.fit(x_train, y_train)
# n_estimators=50: The number of boosting stages to be run. More estimators
# usually improve performance but increase training time.

# learning_rate=1.0: The learning rate shrinks the contribution of each tree.
# There is a trade-off between learning_rate and n_estimators.

# random_state=0: Ensures reproducibility by fixing the
# seed of the random number generator.
print(classification_report(y_val, gbc.predict(x_val), target_names=['Not Fraud', 'Fraud']))

              precision    recall  f1-score   support

   Not Fraud       1.00      1.00      1.00     22771
       Fraud       0.67      0.67      0.67        36

    accuracy                           1.00     22807
   macro avg       0.83      0.83      0.83     22807
weighted avg       1.00      1.00      1.00     22807



In [15]:
# train a linear Support Vector Classifier on the training data
from sklearn.svm import LinearSVC
svc = LinearSVC(class_weight='balanced')
# class_weight='balanced': This parameter adjusts the weights
# inversely proportional to the class frequencies in the input data.
# This is useful for handling class imbalance,
# ensuring that the model pays more attention to the
# minority class (in this case, 'Fraud').

svc.fit(x_train, y_train)
print(classification_report(y_val, svc.predict(x_val), target_names=['Not Fraud', 'Fraud']))

              precision    recall  f1-score   support

   Not Fraud       1.00      1.00      1.00     22771
       Fraud       0.66      0.75      0.70        36

    accuracy                           1.00     22807
   macro avg       0.83      0.87      0.85     22807
weighted avg       1.00      1.00      1.00     22807



/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


In [16]:
frauds = df_2.query('Class == 1')
not_frauds = df_2.query('Class == 0')


the issue we have faced so far is that the dataset is highly imbalanced.
What it does is affect the performance of our model, its like even if the ml model just decides the answer is going to be not fraud, even then it will have 99.8% accuracy which might seem good but takes away the point of detecting the fraud.
When we see how many frauds have been detected its too less, there are so many not frauds that the algos cant differentitate easily between them.

Now we convert it into a balanced dataset. here we randomly pick equal no. of not frauds, and give the algos a 50-50 on frauds and not frauds so there are hgiher stakes for it to understand the difference between them.

Know that this is just to train the thing and might work worse for the whole thing

In [17]:

balanced_df = pd.concat([frauds, not_frauds.sample(len(frauds), random_state=1)])
balanced_df['Class'].value_counts()

Class
1    492
0    492
Name: count, dtype: int64

In [18]:
balanced_df = balanced_df.sample(frac=1, random_state=1)
balanced_df
# randomising it

Time        V1        V2        V3        V4        V5        V6  \
18372   0.170309 -1.762593  0.256143  1.683125 -1.279233 -1.902762  1.004210   
96341   0.380388  1.227614 -0.668974 -0.271785 -0.589440 -0.604795 -0.350285   
248296  0.890522 -0.613696  3.698772 -5.534941  5.620486  1.649263 -2.335145   
264328  0.933932 -0.011624  0.640413  0.868046 -0.505279  0.261938  0.223098   
208904  0.794730 -0.679341  1.217389 -0.316778 -1.086725  0.855349 -0.980760   
...          ...       ...       ...       ...       ...       ...       ...   
81557   0.341393 -4.502731 -3.876484  1.341248  0.113400  0.189428 -0.560985   
276071  0.965803  2.091900 -0.757459 -1.192258 -0.755458 -0.620324 -0.322077   
175971  0.709373  1.972989  0.157281 -1.715078  1.207451  0.681612 -0.615282   
27738   0.200727 -2.439237  2.591458 -2.840126  1.286244 -1.777016 -1.436139   
156988  0.632535  0.745153  2.809299 -5.825406  5.835566  0.512320 -0.615622   

              V7        V8        V9  ...       V21       V22       V23  \
18372  -1.009748 -2.432546  0.458860  ...  2.493579  0.320829 -0.535481   
96341  -0.486365 -0.010809 -0.794944  ... -0.026055 -0.295255 -0.180459   
248296 -0.907188  0.706362 -3.747646  ...  0.319261 -0.471379 -0.075890   
264328  0.239049  0.150877  0.225142  ...  0.069401  0.268024  0.261459   
208904  0.970589  0.133116 -0.357671  ... -0.083048 -0.137032 -0.238920   
...          ...       ...       ...  ...       ...       ...       ...   
81557  -0.140478  0.684651  0.475363  ... -0.140218  0.049411  2.313731   
276071 -1.082511  0.117200 -0.140927  ...  0.288253  0.831939  0.142007   
175971  0.601791 -0.291935 -0.132265  ...  0.098640  0.467533 -0.078973   
27738  -2.206056 -2.282725 -0.292885  ...  1.774460 -0.771390  0.065727   
156988 -2.916576  0.776710 -1.878832  ...  0.284841 -0.874383 -0.083995   

             V24       V25       V26       V27       V28    Amount  Class  
18372   0.499401 -0.915196 -0.423434  0.107049  0.175922  2.906449      0  
96341  -0.436539  0.494649 -0.283738 -0.001128  0.035075  1.062111      1  
248296 -0.667909 -0.642848  0.070600  0.488410  0.292345 -0.307413      1  
264328  0.683742 -1.567901 -0.816674  0.185781  0.283021 -0.272619      0  
208904 -0.617244  0.039020 -0.081848  0.234633  0.128382 -0.307273      0  
...          ...       ...       ...       ...       ...       ...    ...  
81557   0.252330  0.307219  0.859051  0.184033 -0.308269  4.227625      0  
276071  0.592615 -0.196143 -0.136676  0.020182 -0.015470 -0.028645      1  
175971 -0.371882  0.486038 -0.490665 -0.018374 -0.070911  0.075735      0  
27738   0.103916 -0.057578  0.242652 -0.268649 -0.743713  1.443443      1  
156988 -0.651442  0.454594  0.050376  0.756953  0.383869 -0.307413      1  

[984 rows x 31 columns]

In [19]:
balanced_df_np = balanced_df.to_numpy()
# just like earlier we do the ssame stuff on our smaller balanced dataset
x_train_b, y_train_b = balanced_df_np[:700, :-1], balanced_df_np[:700, -1].astype(int)
x_test_b, y_test_b = balanced_df_np[700:842, :-1], balanced_df_np[700:842, -1].astype(int)
x_val_b, y_val_b = balanced_df_np[842:, :-1], balanced_df_np[842:, -1].astype(int)
x_train_b.shape, y_train_b.shape, x_test_b.shape, y_test_b.shape, x_val_b.shape, y_val_b.shape

((700, 30), (700,), (142, 30), (142,), (142, 30), (142,))

In [20]:
pd.Series(y_train_b).value_counts(), pd.Series(y_test_b).value_counts(), pd.Series(y_val_b).value_counts()

(1    353
 0    347
 Name: count, dtype: int64,
 0    73
 1    69
 Name: count, dtype: int64,
 0    72
 1    70
 Name: count, dtype: int64)

In [22]:
from sklearn.linear_model import LogisticRegression
logistic_model_b = LogisticRegression()
logistic_model_b.fit(x_train_b, y_train_b)
logistic_model_b.score(x_train_b, y_train_b)
# same logistic regression model on balanced
# you can notice lesser accuracy but that is because the stakes are high as well

0.9542857142857143

In [23]:
print(classification_report(y_val_b, logistic_model_b.predict(x_val_b), target_names=['Not Fraud', 'Fraud']))

              precision    recall  f1-score   support

   Not Fraud       0.96      0.93      0.94        72
       Fraud       0.93      0.96      0.94        70

    accuracy                           0.94       142
   macro avg       0.94      0.94      0.94       142
weighted avg       0.94      0.94      0.94       142



now the precision and recall are much higher for fraud and the f1 score is great as well but at the same time it affected the not frauds as well compared to earlier. tradeoffs are common

In [24]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import InputLayer, Dense, BatchNormalization
from tensorflow.keras.callbacks import ModelCheckpoint

shallow_nn_b = Sequential()
shallow_nn_b.add(InputLayer((x_train_b.shape[1],)))
shallow_nn_b.add(Dense(2, 'relu'))
shallow_nn_b.add(BatchNormalization())
shallow_nn_b.add(Dense(1, 'sigmoid'))

checkpoint = ModelCheckpoint('shallow_nn_b', save_best_only=True)
shallow_nn_b.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [25]:
shallow_nn_b.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 2)                 62        
                                                                 
 batch_normalization (Batch  (None, 2)                 8         
 Normalization)                                                  
                                                                 
 dense_1 (Dense)             (None, 1)                 3         
                                                                 
Total params: 73 (292.00 Byte)
Trainable params: 69 (276.00 Byte)
Non-trainable params: 4 (16.00 Byte)
_________________________________________________________________


In [27]:
shallow_nn_b.fit(x_train_b, y_train_b, validation_data=(x_val_b, y_val_b), epochs=40, callbacks=checkpoint)

Epoch 1/40
22/22 [==============================] - 1s 35ms/step - loss: 0.2026 - accuracy: 0.9286 - val_loss: 0.1863 - val_accuracy: 0.9366
Epoch 2/40
22/22 [==============================] - 1s 35ms/step - loss: 0.1864 - accuracy: 0.9371 - val_loss: 0.1847 - val_accuracy: 0.9366
Epoch 3/40
22/22 [==============================] - 0s 11ms/step - loss: 0.1805 - accuracy: 0.9329 - val_loss: 0.1851 - val_accuracy: 0.9366
Epoch 4/40
22/22 [==============================] - 0s 8ms/step - loss: 0.1737 - accuracy: 0.9400 - val_loss: 0.1851 - val_accuracy: 0.9366
Epoch 5/40
22/22 [==============================] - 1s 46ms/step - loss: 0.1825 - accuracy: 0.9271 - val_loss: 0.1842 - val_accuracy: 0.9366
Epoch 6/40
22/22 [==============================] - 0s 8ms/step - loss: 0.1850 - accuracy: 0.9300 - val_loss: 0.1876 - val_accuracy: 0.9296
Epoch 7/40
22/22 [==============================] - 0s 8ms/step - loss: 0.1629 - accuracy: 0.9471 - val_loss: 0.1884 - val_accuracy: 0.9366
Epoch 8/40
22/22

In [29]:
def neural_net_predictions(model, x):
  return (model.predict(x).flatten() > 0.5)
neural_net_predictions(shallow_nn_b, x_val)

713/713 [==============================] - 2s 3ms/step


array([0, 0, 0, ..., 0, 0, 0])

In [30]:
print(classification_report(y_val_b, neural_net_predictions(shallow_nn_b, x_val_b), target_names=['Not Fraud', 'Fraud']))

5/5 [==============================] - 0s 2ms/step
              precision    recall  f1-score   support

   Not Fraud       0.94      0.94      0.94        72
       Fraud       0.94      0.94      0.94        70

    accuracy                           0.94       142
   macro avg       0.94      0.94      0.94       142
weighted avg       0.94      0.94      0.94       142



here the result is much better but it feels like it will overfit, so we can make the model a bit less complicated

In [31]:
shallow_nn_b1 = Sequential()
shallow_nn_b1.add(InputLayer((x_train.shape[1],)))
shallow_nn_b1.add(Dense(1, 'relu'))
shallow_nn_b1.add(BatchNormalization())
shallow_nn_b1.add(Dense(1, 'sigmoid'))

checkpoint = ModelCheckpoint('shallow_nn_b1', save_best_only=True)
shallow_nn_b1.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
shallow_nn_b1.fit(x_train_b, y_train_b, validation_data=(x_val_b, y_val_b), epochs=40, callbacks=checkpoint)

Epoch 1/40
22/22 [==============================] - 7s 88ms/step - loss: 0.4656 - accuracy: 0.8043 - val_loss: 0.4349 - val_accuracy: 0.8310
Epoch 2/40
22/22 [==============================] - 1s 68ms/step - loss: 0.4559 - accuracy: 0.8214 - val_loss: 0.4152 - val_accuracy: 0.8592
Epoch 3/40
22/22 [==============================] - 1s 32ms/step - loss: 0.4398 - accuracy: 0.8429 - val_loss: 0.4021 - val_accuracy: 0.8873
Epoch 4/40
22/22 [==============================] - 1s 31ms/step - loss: 0.4283 - accuracy: 0.8329 - val_loss: 0.3940 - val_accuracy: 0.8944
Epoch 5/40
22/22 [==============================] - 1s 32ms/step - loss: 0.4137 - accuracy: 0.8443 - val_loss: 0.3891 - val_accuracy: 0.9014
Epoch 6/40
22/22 [==============================] - 1s 31ms/step - loss: 0.4050 - accuracy: 0.8614 - val_loss: 0.3819 - val_accuracy: 0.9014
Epoch 7/40
22/22 [==============================] - 1s 32ms/step - loss: 0.3873 - accuracy: 0.8643 - val_loss: 0.3782 - val_accuracy: 0.9085
Epoch 8/40
22

In [32]:
shallow_nn_b1.fit(x_train_b, y_train_b, validation_data=(x_val_b, y_val_b), epochs=40, callbacks=checkpoint)

Epoch 1/40
22/22 [==============================] - 0s 12ms/step - loss: 0.1995 - accuracy: 0.9386 - val_loss: 0.2132 - val_accuracy: 0.9155
Epoch 2/40
22/22 [==============================] - 2s 82ms/step - loss: 0.2070 - accuracy: 0.9271 - val_loss: 0.2083 - val_accuracy: 0.9225
Epoch 3/40
22/22 [==============================] - 1s 47ms/step - loss: 0.2080 - accuracy: 0.9400 - val_loss: 0.2071 - val_accuracy: 0.9155
Epoch 4/40
22/22 [==============================] - 0s 8ms/step - loss: 0.2006 - accuracy: 0.9314 - val_loss: 0.2081 - val_accuracy: 0.9085
Epoch 5/40
22/22 [==============================] - 2s 92ms/step - loss: 0.2134 - accuracy: 0.9229 - val_loss: 0.2042 - val_accuracy: 0.9155
Epoch 6/40
22/22 [==============================] - 1s 54ms/step - loss: 0.2086 - accuracy: 0.9214 - val_loss: 0.2013 - val_accuracy: 0.9225
Epoch 7/40
22/22 [==============================] - 0s 9ms/step - loss: 0.2036 - accuracy: 0.9286 - val_loss: 0.2020 - val_accuracy: 0.9225
Epoch 8/40
22/2

In [33]:
print(classification_report(y_val_b, neural_net_predictions(shallow_nn_b1, x_val_b), target_names=['Not Fraud', 'Fraud']))

5/5 [==============================] - 0s 4ms/step
              precision    recall  f1-score   support

   Not Fraud       0.93      0.93      0.93        72
       Fraud       0.93      0.93      0.93        70

    accuracy                           0.93       142
   macro avg       0.93      0.93      0.93       142
weighted avg       0.93      0.93      0.93       142



In [36]:

from sklearn.ensemble import RandomForestClassifier

rf_b = RandomForestClassifier(max_depth=2, n_jobs=-1)
rf_b.fit(x_train_b, y_train_b)
print(classification_report(y_val_b, rf_b.predict(x_val_b), target_names=['Not Fraud', 'Fraud']))


              precision    recall  f1-score   support

   Not Fraud       0.92      1.00      0.96        72
       Fraud       1.00      0.91      0.96        70

    accuracy                           0.96       142
   macro avg       0.96      0.96      0.96       142
weighted avg       0.96      0.96      0.96       142



In [38]:
from sklearn.ensemble import GradientBoostingClassifier

gbc_b = GradientBoostingClassifier(n_estimators=50, learning_rate=1.0, max_depth=2, random_state=0)
gbc_b.fit(x_train_b, y_train_b)
print(classification_report(y_val_b, gbc_b.predict(x_val_b), target_names=['Not Fraud', 'Fraud']))

              precision    recall  f1-score   support

   Not Fraud       0.94      0.92      0.93        72
       Fraud       0.92      0.94      0.93        70

    accuracy                           0.93       142
   macro avg       0.93      0.93      0.93       142
weighted avg       0.93      0.93      0.93       142



In [39]:
from sklearn.svm import LinearSVC
svc_b = LinearSVC(class_weight='balanced')
svc_b.fit(x_train_b, y_train_b)
print(classification_report(y_val_b, svc_b.predict(x_val_b), target_names=['Not Fraud', 'Fraud']))

              precision    recall  f1-score   support

   Not Fraud       0.96      0.93      0.94        72
       Fraud       0.93      0.96      0.94        70

    accuracy                           0.94       142
   macro avg       0.94      0.94      0.94       142
weighted avg       0.94      0.94      0.94       142



/usr/local/lib/python3.10/dist-packages/sklearn/svm/_base.py:1244: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


ok so out of all them, i got the best result for the random forest one with 0.96 everything
We can try more randomly in terms of not frauds, or try including more not frauds aand see improvements, it might help with the overfitting further
i believe some concrete math might be required to handle the trade offs

Either ways, this is a good lesson on imbalanced datasets